index.d.ts

```diff
@@ -1,184 +1,261 @@
-import { Dispatch, SetStateAction, ReactNode } from 'react';
-
 declare module 'react-table' {
-    export interface Cell<D> {
-        render: (type: string) => any;
-        getCellProps: () => any;
-        column: Column<D>;
-        row: Row<D>;
-        state: any;
-        value: any;
-    }
+  import { ReactNode, ReactText, Dispatch, SetStateAction } from 'react'
+    export type SortingRule<D> = {
```

Comment:
What do y'all think about adding a second generic for the key? I find it useful even if it's a bit of a PITA to setup initially. 

`SortingRule<D, K extends Extract<keyof D, 'string'> = never>` 

The `never` just means that this one should always get the second generic. Alternatively, we can default it to `Extract<keyof D, 'string'>` which makes it optional.

Also, NOTE, the `Extract<keyof D, 'string'>` enforces that the keys that we use are strings. I have seen numbers have issues when used as keys.

Reply:
I like the idea, but I've had so much trouble actually getting it to work.

e.g. I *think* that this should work: 

```ts
type SortingRule<D, K extends Extract<keyof D, 'string'> = never> = {
    id: K
    desc: boolean
  }

type User = {
  name: string
  email: string
}

const DEFAULT_SORT: SortingRule<User>[] = [
  {
    id: 'email',  // <- error here: Type 'string' is not assignable to type 'never'
    desc: false
  }
]
```

see: http://www.typescriptlang.org/play/#code/C4TwDgpgBAyg9gJ2ASwHYHMBKBXANhAHgBEAaKAaSggA9gJUATAZygFFaEBDAY2AIGsIIOADMopKAHImwBGnSSAfFAC8UVBABuEBMrUBvAFBQTUZAwBcFY6YYQm3KwCM4cfJ1Q2AvocOhIUACqTDqqUEYmqJwAthBWMnIYNhDRnMi48bLyhj6G3HCoMuKsAGIAgoEAMgAqAPowAPKY1VbwSPI4+ATBOooA2gC6YX02EabmVpIpabiSJDYmdg5WIpy4Id6GA4ZAA

Reply:
Ah sorry, I was too lazy to look up the syntax last night and I used `Extract` incorrectly. In the example you provided, 

`type SortingRule<D, K extends Extract<keyof D, 'string'> = never> = {` 

should be 

`type SortingRule<D, K extends Extract<keyof D, string> = never> = {`

That fixes the issue with it always being `never`.

The second issue in the example you provided is you are using the typing incorrectly. The `= never` portion means that you MUST provide the second argument.

I personally, prefer the simplicity of it handling the key by default. so the rule would go from:

`type SortingRule<D, K extends Extract<keyof D, string> = never> = {`

to

`type SortingRule<D, K extends Extract<keyof D, string> = Extract<keyof D, string>> = {`

which works in the provided typescript playground you provided.

Reply:
On that's very cool, I didn't know you could do that.

That said, on thinking about it a bit more I wonder if this isn't the wrong approach.

Because it suits my own use-case so well, I've tended to focus on the accessor being either a valid key of the type, or a function returning a value, in which case id is the key.

The problem is that that's not really the whole story.

From the docs:

>accessor: String | Function
>...
>If a string is passed, the column's value will be looked up on the original row via that key, eg. If your column's accessor is firstName then its value would be read from row['firstName']. **You can also specify deeply nested values with accessors like info.hobbies or even address[0].street**

I don't know how we can type that last part other than using string.

Thoughts?

Reply:
I suppose that this comment mostly applies to accessor, and I see now that you were suggesting accessor as K | string | function - which I guess improves the intellisense, though it doesn't do as much for the type checking.

Reply:
Correct. So this comment definitely applies more so to the `accessor` but can be useful for keeping the key type abstracted for multiple uses and keeping the string extraction logic in one place in the type.